### PR TITLE
kz_json_tests: cleanup JSON types

### PR DESCRIPTION
**Pull request by jamesaimonetti**
proper was complaining of no_base_case for some types. Introducing the
appropriate base cases (like empty list) has resolved those complaints.

### CODE REVIEW

File: core/kazoo_json/include/kazoo_json.hrl [hunk 1]
```diff
@@ -13,42 +13,36 @@
...
 -define(EMPTY_JSON_OBJECT, ?JSON_WRAPPER([])).
 
--type object() :: ?JSON_WRAPPER(json_proplist()) | ?EMPTY_JSON_OBJECT.
--type objects() :: [object()].
-
--type json_string() :: ne_binary() | atom().
--type json_strings() :: [json_string()].
+-type json_value() :: 'null'
+                    | 'true'
+                    | 'false'
```
Comment (fenollp) [inline core/kazoo_json/include/kazoo_json.hrl]: `boolean()`

Comment (jamesaimonetti) [inline core/kazoo_json/include/kazoo_json.hrl]: True, I lifted these from jiffy

File: core/kazoo_json/include/kazoo_json.hrl [hunk 2]
```diff
@@ -13,42 +13,35 @@
...
-
--type json_string() :: ne_binary() | atom().
--type json_strings() :: [json_string()].
+-type json_value() :: 'null'
+                    | boolean()
+                    | json_string()
+                    | json_number()
+                    | object()
+                    | json_array().
+-type json_values() :: [json_value()] | [].
```
Comment (fenollp) [inline core/kazoo_json/include/kazoo_json.hrl]: This is exactly the same as `[json_value()]`.

File: core/kazoo_json/include/kazoo_json.hrl [hunk 3]
```diff
@@ -13,42 +13,35 @@
...
+
+-type json_array()  :: [json_value()].
+-type json_string() :: atom() | binary().
 -type json_number() :: integer() | float().
--type json_array() :: [json_term()].
 
--type key() :: json_string() | pos_integer().
+-type object() :: {json_proplist()} | ?EMPTY_JSON_OBJECT.
+-type objects() :: [object()] | [].
```
Comment (fenollp) [inline core/kazoo_json/include/kazoo_json.hrl]: This is exactly the same as `[object()]`.

File: core/kazoo_json/include/kazoo_json.hrl [hunk 4]
```diff
@@ -13,42 +13,35 @@
...
+                    | json_array().
+-type json_values() :: [json_value()] | [].
+
+-type json_array()  :: [json_value()].
+-type json_string() :: atom() | binary().
 -type json_number() :: integer() | float().
--type json_array() :: [json_term()].
 
--type key() :: json_string() | pos_integer().
+-type object() :: {json_proplist()} | ?EMPTY_JSON_OBJECT.
```
Comment (fenollp) [inline core/kazoo_json/include/kazoo_json.hrl]: What are the actual changes here? Why move stuff around?

File: core/kazoo_json/src/kz_json.erl [hunk 5]
```diff
@@ -94,8 +94,7 @@
 -include_lib("kazoo/include/kz_log.hrl").
 -include_lib("kazoo_json/include/kazoo_json.hrl").
 
--export_type([json_term/0, api_json_term/0, json_terms/0
```
Comment (fenollp) [inline core/kazoo_json/src/kz_json.erl]: At least `api_json_term/0` is used throughout Kazoo. Why remove it?

File: core/kazoo_json/include/kazoo_json.hrl [hunk 6]
```diff
@@ -13,42 +13,35 @@
...
 
--type json_proplist_key() :: path().
--type json_proplist_kv(K, V) :: [{K, V}].
--type json_proplist_k(K) :: json_proplist_kv(K, json_term()).
--type json_proplist() :: json_proplist_kv(json_proplist_key(), json_term()).
--type json_proplists() :: [json_proplist(),...].
-
--type json_iolist() :: {'json', iolist()}.
--type json_term() :: boolean()
-                   | json_string() | <<>>
```
Comment (fenollp) [inline core/kazoo_json/include/kazoo_json.hrl]: No more empty binary in JSON values?

File: core/kazoo_json/include/kazoo_json.hrl [hunk 7]
```diff
@@ -13,42 +13,35 @@
 
 -define(EMPTY_JSON_OBJECT, ?JSON_WRAPPER([])).
 
--type object() :: ?JSON_WRAPPER(json_proplist()) | ?EMPTY_JSON_OBJECT.
--type objects() :: [object()].
-
--type json_string() :: ne_binary() | atom().
--type json_strings() :: [json_string()].
+-type json_value() :: 'null'
```
Comment (fenollp) [inline core/kazoo_json/include/kazoo_json.hrl]: Why the rename?

File: core/kazoo_json/src/kz_json.erl [hunk 8]
```diff
@@ -680,8 +679,8 @@ find_value(Key, Value, [JObj|JObjs], Default) ->
         _Value -> find_value(Key, Value, JObjs, Default)
     end.
 
--spec get_first_defined(paths(), object()) -> api_json_term().
--spec get_first_defined(paths(), object(), Default) -> Default | json_term().
+-spec get_first_defined(paths(), object()) -> json_value() | 'undefined'.
```
Comment (fenollp) [inline core/kazoo_json/src/kz_json.erl]: `api_json_term()`

File: core/kazoo_json/test/kz_json_test.erl [hunk 9]
```diff
@@ -55,18 +56,15 @@ prop_get_value() ->
                       end)
            ).
 
--define(DO_NOT_RUN_QC_SET_VALUE, true).
--ifndef(DO_NOT_RUN_QC_SET_VALUE).
```
Comment (fenollp) [inline core/kazoo_json/test/kz_json_test.erl]: This test was disabled because it sometimes failed in CI. How are you fixing it?

File: core/kazoo_json/include/kazoo_json.hrl (hunk 2 above)
Comment (jamesaimonetti) [inline core/kazoo_json/include/kazoo_json.hrl]: PropEr was complaining about no base case for the type. This shut it up

File: core/kazoo_json/include/kazoo_json.hrl (hunk 3 above)
Comment (jamesaimonetti) [inline core/kazoo_json/include/kazoo_json.hrl]: Same here

File: core/kazoo_json/test/kz_json_test.erl (hunk 9 above)
Comment (jamesaimonetti) [inline core/kazoo_json/test/kz_json_test.erl]: I have yet to see it fail with these changes. Disabling tests is wrong and bad and you should feel bad for doing it! :)

File: core/kazoo_json/src/kz_json.erl (hunk 5 above)
Comment (fenollp) [inline core/kazoo_json/src/kz_json.erl]: Let me reformulate: you renamed a supertype. Why did you remove some subtypes though?

File: core/kazoo_json/include/kazoo_json.hrl (hunk 7 above)
Comment (fenollp) [inline core/kazoo_json/include/kazoo_json.hrl]: Could you revert the renaming? It hides actual changes and you didn't propagate it throughout the codebase.

File: core/kazoo_json/test/kz_json_test.erl (hunk 9 above)
Comment (fenollp) [inline core/kazoo_json/test/kz_json_test.erl]: Right

Comment (fenollp) [inline core/kazoo_json/test/kz_json_test.erl]: Could you fix this one too? https://travis-ci.org/2600hz/kazoo/jobs/171549029#L2398

**Comment by lazedo** [comment]
@fenollp what changes are you requesting here ?

File: core/kazoo_json/src/kz_json.erl (hunk 5 above)
Comment (fenollp) [inline core/kazoo_json/src/kz_json.erl]: These ^ are the changes I'm requesting.

**Comment by jamesaimonetti** [comment]
@fenollp @lazedo this is ready (addressed the dialyzer unmatched return).
